README.md
<p align="center">
  <img src="./src/qt/res/images/donu-logo.png" alt="Donito, Inc.">
</p>

## The Donu Core Wallet

The Donu Core wallet gives people the ability to send, receive and stake DONU as well as run masternodes on the Donu blockchain.  It also allows users of [The Donito Platform](https://www.doni.to) to enjoy benefits such as reduced platform fees, and other applicable discounts and future rewards.  Additional wallets such as light wallets, mobile and other implementations will be made available as well with the end goal being to completely decentralize the Donu platform.

## About Donito and Donu

Donito (pronounced: dough-knee-toe) is dervied from an Esperanto word meaning "to give", which is the central theme for the project. The ultimate goal of Donito is to facilitate an easy way for creators/charities/workers to accept a large selection of cryptocurrencies from patrons.  Donito also has a buy-back program, in which all fees generated by the platform are used to buy DONU from the open market, which are optionally burned in an effort to prevent over-supply.

You can learn more about Donito by visiting [The Donito Platform Website](https://www.doni.to) or [The Donu Crypto Website](https://www.donu.io).

## Blockchain/DONU specifications

**Block Time** 1 Minute  
**Block Finder Reward** 0.4 DONU  
**Masternode Winner Reward** 0.4 DONU  
**Platform Emissions** 0.2 DONU  
**Network Port** 26380  
**Default RPC Port** 36380  
**Masternode collateral** 5000 DONU  
**Minimum input size for staking** 500 DONU  
**Minimum stake input age** 30 days  
**Maximum stake input age** 60 days

The first 86400 blocks or approximately 60 days worth of blocks will be mined by the team to mature inputs sent out during the swap and others transacted during PoW for the PoS phase.
PoW is enabled only in the event PoS needs to be stopped for emergency maintenance, PoS has any issues or in the case of a vulnerability and thus the PoW block reward
goes to a platform address.

This serves dampen monetary incentive for anyone looking to ramp up the difficulty as it would yield no return.

You can use [The DONU Block Explorer](https://chainz.cryptoid.info/donu) to look up transactions, get block information and other valuable Donu blockchain and network details.

## Getting help

You can get help by visiting [our support portal & knowledgebase](https://www.donu.support) or by joining us through any of our communication mediums:

- [Telegram](https://www.donu.io/telegram)
- [Discord](https://www.donu.io/discord)
- [Matrix Chat](https://www.donu.io/matrix)
- [Twitter](https://www.donu.io/twitter)
- [Facebook](https://www.donu.io/facebook)
- [Reddit](https://www.donu.io/reddit)

## History behind Donu and the team

Before it was "Donu", there was Neos.  Neos was a project created back in 2014 with the intention of bringing a unique product with quality support and innovation into the industry.  Once hearing about the issues people were having with Patreon and the over-policing of social media, we started looking into this sector to learn more and see what gaps could be filled.  Discovering the incredibly large fees charged by most services and solutions alongside the lack of unbias to service members we decided to build Donu.


debian/changelog
donu (0.17.1-bionic1) bionic; urgency=high

  * First Release

 -- Ahmed Bodiwala (ahmedbodi) <admin@multicoin.co>  Thu, 11 Jul 2019 12:41:00 +0000

bitcoin (0.16.2-trusty4) trusty; urgency=medium

  * Fix set direction.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 14 Aug 2018 14:37:00 -0400

bitcoin (0.16.2-trusty3) trusty; urgency=medium

  * Ignore errors creating the bitcoin user in package configure.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 14 Aug 2018 09:36:00 -0400

bitcoin (0.16.2-trusty2) trusty; urgency=medium

  * Use qt5 on bionic.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 30 Jul 2018 16:46:00 -0400

bitcoin (0.16.2-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 30 Jul 2018 12:34:00 -0400

bitcoin (0.16.0-trusty2) trusty; urgency=medium

  * Add systemd service to bitcoind

 -- Thomas M Steenholdt <tsteenholdt@cascadetechnologypartners.com>  Wed, 18 Apr 2018 16:40:00 -0200

bitcoin (0.16.0-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 05 Mar 2018 11:18:00 -0500

bitcoin (0.15.1-trusty1) trusty; urgency=medium

  * New upstream release.
  * Remove backported patches (which are now upstream).

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Sat, 11 Nov 2017 16:51:00 -0500

bitcoin (0.15.0-artful9) artful; urgency=medium

  * Mark for artful.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Sat, 21 Oct 2017 20:56:00 -0400

bitcoin (0.15.0-trusty9) trusty; urgency=medium

  * Add missing xvfb dep.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 14 Sep 2017 22:47:00 -0400

bitcoin (0.15.0-trusty8) trusty; urgency=medium

  * Backport #11332 to fix Qt settings upgrade segfault.
  * Use qt5 on arm to fix Qt test segfault.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 14 Sep 2017 22:03:00 -0400

bitcoin (0.15.0-trusty7) trusty; urgency=medium

  * Backport #11210 to fix build inside launchpad.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 12 Sep 2017 16:13:00 -0400

bitcoin (0.15.0-trusty6) trusty; urgency=medium

  * Make launchpad print more debug information.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 12 Sep 2017 12:50:00 -0400

bitcoin (0.15.0-trusty5) trusty; urgency=medium

  * Use proper makefile comparison to fix Xvfb start.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 12 Sep 2017 12:49:00 -0400

bitcoin (0.15.0-trusty4) trusty; urgency=medium

  * Use full path for start-stop-daemon to fix Xvfb start.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 12 Sep 2017 11:44:00 -0400

bitcoin (0.15.0-trusty3) trusty; urgency=medium

  * Fix DISPLAY setting when using xvfb.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 11 Sep 2017 20:06:00 -0400

bitcoin (0.15.0-trusty2) trusty; urgency=medium

  * Use xvfb to run qt4 tests.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 11 Sep 2017 17:31:00 -0400

bitcoin (0.15.0-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 11 Sep 2017 16:17:00 -0400

bitcoin (0.14.2-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Fri, 23 Jun 2017 18:21:00 -0400

bitcoin (0.14.1-trusty4) trusty; urgency=medium

  * Re-enable UPnP support.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Fri, 05 May 2017 13:28:00 -0400

bitcoin (0.14.1-trusty3) trusty; urgency=medium

  * Build with qt5 if we are on a non-Ubuntu (ie non-Unity) distro.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 04 May 2017 17:13:00 -0400

bitcoin (0.14.1-trusty2) trusty; urgency=medium

  * Bump minimum boost version in deps.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 04 May 2017 17:12:00 -0400

bitcoin (0.14.1-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Sat, 22 Apr 2017 17:10:00 -0400

bitcoin (0.14.0-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Wed, 08 Mar 2017 10:30:00 -0500

bitcoin (0.13.2-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 05 Jan 2017 09:59:00 -0500

bitcoin (0.13.1-trusty2) trusty; urgency=medium

  * Revert to Qt4, due to https://github.com/bitcoin/bitcoin/issues/9038

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 31 Oct 2016 11:16:00 -0400

bitcoin (0.13.1-trusty1) trusty; urgency=medium

  * New upstream release.
  * Backport updated bitcoin-qt.desktop from upstream master
  * Add zmq dependency
  * Switch to Qt5 (breaks precise, but that was already broken by C++11)

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Thu, 27 Oct 2016 17:32:00 -0400

bitcoin (0.13.0-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Sun, 04 Sep 2016 22:09:00 -0400

bitcoin (0.12.1-trusty1) trusty; urgency=medium

  * New upstream release.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Mon, 18 Apr 2016 14:26:00 -0700

bitcoin (0.12.0-trusty6) trusty; urgency=medium

  * Fix program-options dep.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Fri, 25 Mar 2016 21:41:00 -0700

bitcoin (0.12.0-trusty5) trusty; urgency=medium

  * Test explicit --with-gui

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 23 Feb 2015 23:25:00 -0800

bitcoin (0.12.0-trusty4) trusty; urgency=medium

  * Fix libevent-dev dep.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 23 Feb 2015 23:25:00 -0800

bitcoin (0.12.0-trusty3) trusty; urgency=medium

  * Fix precise boost dep.

 -- Matt Corallo (BlueMatt) <matt@mattcorallo.com>  Tue, 23 Feb 2015 19:55:00 -0800

bitcoin (0.12.0-trusty2) trusty; urgency=medium

  * Fix libevent dep.


src/hash.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2018 The Bitcoin Core Developers
// Copyright (C) 2014-2019 The Dash Core Developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#ifndef BITCOIN_HASH_H
#define BITCOIN_HASH_H

#include <crypto/ripemd160.h>
#include <crypto/sha256.h>
#include "crypto/sph_blake.h"
#include "crypto/sph_bmw.h"
#include "crypto/sph_groestl.h"
#include "crypto/sph_jh.h"
#include "crypto/sph_keccak.h"
#include "crypto/sph_skein.h"
#include "crypto/sph_luffa.h"
#include "crypto/sph_cubehash.h"
#include "crypto/sph_shavite.h"
#include "crypto/sph_simd.h"
#include "crypto/sph_echo.h"
#include "crypto/sph_hamsi.h"
#include "crypto/sph_fugue.h"
#include "crypto/sph_shabal.h"
#include "crypto/sph_whirlpool.h"
#include "crypto/sph_sha2.h"
#include "crypto/sph_haval.h"
#include "crypto/sph_tiger.h"
#include "crypto/lyra2.h"
#include "crypto/gost_streebog.h"
#include "crypto/SWIFFTX/SWIFFTX.h"
#include <prevector.h>
#include <serialize.h>
#include <uint256.h>
#include <version.h>

#include <vector>

typedef uint256 ChainCode;

/** A hasher class for DONU's 256-bit hash (double SHA-256). */
class CHash256 {
private:
    CSHA256 sha;
public:
    static const size_t OUTPUT_SIZE = CSHA256::OUTPUT_SIZE;

    void Finalize(unsigned char hash[OUTPUT_SIZE]) {
        unsigned char buf[CSHA256::OUTPUT_SIZE];
        sha.Finalize(buf);
        sha.Reset().Write(buf, CSHA256::OUTPUT_SIZE).Finalize(hash);
    }

    CHash256& Write(const unsigned char *data, size_t len) {
        sha.Write(data, len);
        return *this;
    }

    CHash256& Reset() {
        sha.Reset();
        return *this;
    }
};

/** A hasher class for DONU's 160-bit hash (SHA-256 + RIPEMD-160). */
class CHash160 {
private:
    CSHA256 sha;
public:
    static const size_t OUTPUT_SIZE = CRIPEMD160::OUTPUT_SIZE;

    void Finalize(unsigned char hash[OUTPUT_SIZE]) {
        unsigned char buf[CSHA256::OUTPUT_SIZE];
        sha.Finalize(buf);
        CRIPEMD160().Write(buf, CSHA256::OUTPUT_SIZE).Finalize(hash);
    }

    CHash160& Write(const unsigned char *data, size_t len) {
        sha.Write(data, len);
        return *this;
    }

    CHash160& Reset() {
        sha.Reset();
        return *this;
    }
};

/** Compute the 256-bit hash of an object. */
template<typename T1>
inline uint256 Hash(const T1 pbegin, const T1 pend)
{
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(pbegin == pend ? pblank : (const unsigned char*)&pbegin[0], (pend - pbegin) * sizeof(pbegin[0]))
              .Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 256-bit hash of the concatenation of two objects. */
template<typename T1, typename T2>
inline uint256 Hash(const T1 p1begin, const T1 p1end,
                    const T2 p2begin, const T2 p2end) {
    static const unsigned char pblank[1] = {};
    uint256 result;
    CHash256().Write(p1begin == p1end ? pblank : (const unsigned char*)&p1begin[0], (p1end - p1begin) * sizeof(p1begin[0]))
              .Write(p2begin == p2end ? pblank : (const unsigned char*)&p2begin[0], (p2end - p2begin) * sizeof(p2begin[0]))
              .Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 160-bit hash an object. */
template<typename T1>
inline uint160 Hash160(const T1 pbegin, const T1 pend)
{
    static unsigned char pblank[1] = {};
    uint160 result;
    CHash160().Write(pbegin == pend ? pblank : (const unsigned char*)&pbegin[0], (pend - pbegin) * sizeof(pbegin[0]))
              .Finalize((unsigned char*)&result);
    return result;
}

/** Compute the 160-bit hash of a vector. */
inline uint160 Hash160(const std::vector<unsigned char>& vch)
{
    return Hash160(vch.begin(), vch.end());
}

/** Compute the 160-bit hash of a vector. */
template<unsigned int N>
inline uint160 Hash160(const prevector<N, unsigned char>& vch)
{
    return Hash160(vch.begin(), vch.end());
}

/** A writer stream (for serialization) that computes a 256-bit hash. */
class CHashWriter
{
private:
    CHash256 ctx;

    const int nType;
    const int nVersion;
public:

    CHashWriter(int nTypeIn, int nVersionIn) : nType(nTypeIn), nVersion(nVersionIn) {}

    int GetType() const { return nType; }
    int GetVersion() const { return nVersion; }

    void write(const char *pch, size_t size) {
        ctx.Write((const unsigned char*)pch, size);
    }

    // invalidates the object
    uint256 GetHash() {
        uint256 result;
        ctx.Finalize((unsigned char*)&result);
        return result;
    }

    template<typename T>
    CHashWriter& operator<<(const T& obj) {
        // Serialize to this stream
        ::Serialize(*this, obj);
        return (*this);
    }
};

/** Reads data from an underlying stream, while hashing the read data. */
template<typename Source>
class CHashVerifier : public CHashWriter
{
private:
    Source* source;

public:
    explicit CHashVerifier(Source* source_) : CHashWriter(source_->GetType(), source_->GetVersion()), source(source_) {}

    void read(char* pch, size_t nSize)
    {
        source->read(pch, nSize);
        this->write(pch, nSize);
    }

    void ignore(size_t nSize)
    {
        char data[1024];
        while (nSize > 0) {
            size_t now = std::min<size_t>(nSize, 1024);
            read(data, now);
            nSize -= now;
        }
    }

    template<typename T>
    CHashVerifier<Source>& operator>>(T&& obj)
    {
        // Unserialize from this stream
        ::Unserialize(*this, obj);
        return (*this);
    }
};

/** Compute the 256-bit hash of an object's serialization. */
template<typename T>
uint256 SerializeHash(const T& obj, int nType=SER_GETHASH, int nVersion=PROTOCOL_VERSION)
{
    CHashWriter ss(nType, nVersion);
    ss << obj;
    return ss.GetHash();
}

unsigned int MurmurHash3(unsigned int nHashSeed, const std::vector<unsigned char>& vDataToHash);

void BIP32Hash(const ChainCode &chainCode, unsigned int nChild, unsigned char header, const unsigned char data[32], unsigned char output[64]);

/** SipHash-2-4 */
class CSipHasher
{
private:
    uint64_t v[4];
    uint64_t tmp;
    int count;

public:
    /** Construct a SipHash calculator initialized with 128-bit key (k0, k1) */
    CSipHasher(uint64_t k0, uint64_t k1);
    /** Hash a 64-bit integer worth of data
     *  It is treated as if this was the little-endian interpretation of 8 bytes.
     *  This function can only be used when a multiple of 8 bytes have been written so far.
     */
    CSipHasher& Write(uint64_t data);
    /** Hash arbitrary bytes. */
    CSipHasher& Write(const unsigned char* data, size_t size);
    /** Compute the 64-bit SipHash-2-4 of the data written so far. The object remains untouched. */
    uint64_t Finalize() const;
};

/** Optimized SipHash-2-4 implementation for uint256.
 *
 *  It is identical to:
 *    SipHasher(k0, k1)
 *      .Write(val.GetUint64(0))
 *      .Write(val.GetUint64(1))
 *      .Write(val.GetUint64(2))
 *      .Write(val.GetUint64(3))
 *      .Finalize()
 */
uint64_t SipHashUint256(uint64_t k0, uint64_t k1, const uint256& val);
uint64_t SipHashUint256Extra(uint64_t k0, uint64_t k1, const uint256& val, uint32_t extra);

/* x22i-hash */
template<typename T1>
inline uint256 HashX22I(const T1 pbegin, const T1 pend)
{
    sph_blake512_context      ctx_blake;
    sph_bmw512_context        ctx_bmw;
    sph_groestl512_context    ctx_groestl;
    sph_jh512_context         ctx_jh;
    sph_keccak512_context     ctx_keccak;
    sph_skein512_context      ctx_skein;
    sph_luffa512_context      ctx_luffa;
    sph_cubehash512_context   ctx_cubehash;
    sph_shavite512_context    ctx_shavite;
    sph_simd512_context       ctx_simd;
    sph_echo512_context       ctx_echo;
    sph_hamsi512_context      ctx_hamsi;
    sph_fugue512_context      ctx_fugue;
    sph_shabal512_context     ctx_shabal;
    sph_whirlpool_context     ctx_whirlpool;
    sph_sha512_context        ctx_sha2;
    sph_haval256_5_context    ctx_haval;
    sph_tiger_context         ctx_tiger;
    sph_gost512_context       ctx_gost;
    sph_sha256_context        ctx_sha;
    static unsigned char pblank[1];
    uint512 hash[22];
	
    sph_blake512_init(&ctx_blake);
    sph_blake512 (&ctx_blake, (pbegin == pend ? pblank : static_cast<const void*>(&pbegin[0])), (pend - pbegin) * sizeof(pbegin[0]));
    sph_blake512_close(&ctx_blake, static_cast<void*>(&hash[0]));
	
    sph_bmw512_init(&ctx_bmw);
    sph_bmw512 (&ctx_bmw, static_cast<const void*>(&hash[0]), 64);
    sph_bmw512_close(&ctx_bmw, static_cast<void*>(&hash[1]));
	
    sph_groestl512_init(&ctx_groestl);
	sph_groestl512 (&ctx_groestl, static_cast<const void*>(&hash[1]), 64);
    sph_groestl512_close(&ctx_groestl, static_cast<void*>(&hash[2]));
	
    sph_skein512_init(&ctx_skein);
    sph_skein512 (&ctx_skein, static_cast<const void*>(&hash[2]), 64);
    sph_skein512_close(&ctx_skein, static_cast<void*>(&hash[3]));
	
    sph_jh512_init(&ctx_jh);
    sph_jh512 (&ctx_jh, static_cast<const void*>(&hash[3]), 64);
    sph_jh512_close(&ctx_jh, static_cast<void*>(&hash[4]));
	
	sph_keccak512_init(&ctx_keccak);
    sph_keccak512 (&ctx_keccak, static_cast<const void*>(&hash[4]), 64);
    sph_keccak512_close(&ctx_keccak, static_cast<void*>(&hash[5]));
	
    sph_luffa512_init(&ctx_luffa);
    sph_luffa512 (&ctx_luffa, static_cast<void*>(&hash[5]), 64);
    sph_luffa512_close(&ctx_luffa, static_cast<void*>(&hash[6]));
	
    sph_cubehash512_init(&ctx_cubehash);
    sph_cubehash512 (&ctx_cubehash, static_cast<const void*>(&hash[6]), 64);
    sph_cubehash512_close(&ctx_cubehash, static_cast<void*>(&hash[7]));
	
    sph_shavite512_init(&ctx_shavite);
    sph_shavite512(&ctx_shavite, static_cast<const void*>(&hash[7]), 64);
    sph_shavite512_close(&ctx_shavite, static_cast<void*>(&hash[8]));
	
    sph_simd512_init(&ctx_simd);
    sph_simd512 (&ctx_simd, static_cast<const void*>(&hash[8]), 64);
    sph_simd512_close(&ctx_simd, static_cast<void*>(&hash[9]));
	
    sph_echo512_init(&ctx_echo);
    sph_echo512 (&ctx_echo, static_cast<const void*>(&hash[9]), 64);
    sph_echo512_close(&ctx_echo, static_cast<void*>(&hash[10]));
	
    sph_hamsi512_init(&ctx_hamsi);
    sph_hamsi512 (&ctx_hamsi, static_cast<const void*>(&hash[10]), 64);
    sph_hamsi512_close(&ctx_hamsi, static_cast<void*>(&hash[11]));
	
    sph_fugue512_init(&ctx_fugue);
    sph_fugue512 (&ctx_fugue, static_cast<const void*>(&hash[11]), 64);
    sph_fugue512_close(&ctx_fugue, static_cast<void*>(&hash[12]));
	
    sph_shabal512_init(&ctx_shabal);
    sph_shabal512 (&ctx_shabal, static_cast<const void*>(&hash[12]), 64);
    sph_shabal512_close(&ctx_shabal, static_cast<void*>(&hash[13]));
	
    sph_whirlpool_init(&ctx_whirlpool);
    sph_whirlpool (&ctx_whirlpool, static_cast<const void*>(&hash[13]), 64);
    sph_whirlpool_close(&ctx_whirlpool, static_cast<void*>(&hash[14]));
	
    sph_sha512_init(&ctx_sha2);
    sph_sha512 (&ctx_sha2, static_cast<const void*>(&hash[14]), 64);
    sph_sha512_close(&ctx_sha2, static_cast<void*>(&hash[15]));
	
    unsigned char temp[SWIFFTX_OUTPUT_BLOCK_SIZE] = {0};
    InitializeSWIFFTX();
    ComputeSingleSWIFFTX((unsigned char*)&hash[12], temp, false);
	
    memcpy((unsigned char*)&hash[16], temp, 64);
    sph_haval256_5_init(&ctx_haval);
    sph_haval256_5 (&ctx_haval, static_cast<const void*>(&hash[16]), 64);
    sph_haval256_5_close(&ctx_haval, static_cast<void*>(&hash[17]));
	
    sph_tiger_init(&ctx_tiger);
    sph_tiger (&ctx_tiger, static_cast<const void*>(&hash[17]), 64);
    sph_tiger_close(&ctx_tiger, static_cast<void*>(&hash[18]));
	
    LYRA2(static_cast<void*>(&hash[19]), 32, static_cast<const void*>(&hash[18]), 32, static_cast<const void*>(&hash[18]), 32, 1, 4, 4);
    
    sph_gost512_init(&ctx_gost);
    sph_gost512 (&ctx_gost, static_cast<const void*>(&hash[19]), 64);
    sph_gost512_close(&ctx_gost, static_cast<void*>(&hash[20]));
	
    sph_sha256_init(&ctx_sha);
    sph_sha256 (&ctx_sha, static_cast<const void*>(&hash[20]), 64);
    sph_sha256_close(&ctx_sha, static_cast<void*>(&hash[21]));
	
    return hash[21].trim256();
}

#endif // BITCOIN_HASH_H
